File: src/agents/alerter/src/extraction/extractor.js
'use strict';

// Phase 38 Plan 03 Task 2: extractor.js -- the heart of the extraction pipeline.
// Phase 39 D-03: farmerCorrection added to buildInitialUserContent for the EDIT loop.
//
// extract({captures, inFlightDraft}):
//   1. Build content blocks via multimodal.buildContentBlocks. Multimodal fusion
//      (EXT-03) happens here: text + transcript + image_paths all go in ONE call.
//   2. Build the Anthropic request with cached system + few-shot + a forced tool_use:
//      tools=[submit_extraction], tool_choice={type:'tool',name:'submit_extraction'}.
//   3. Call client.messages.create.
//   4. Find the tool_use block; Zod safeParse via validator.validateDraft against
//      the Submission schema.
//   5. On Zod failure (first attempt), send back a tool_result with is_error:true
//      so the model can fix its call. One retry max. Second failure returns
//      {ok:false, reason:'schema_invalid'}.
//   6. Never throws. SDK errors surface as {ok:false, reason}.
//
// V2: ANTHROPIC_API_KEY only crosses into `new Anthropic({ apiKey })`; never logged.

const crypto = require('crypto');
const Anthropic = require('@anthropic-ai/sdk');
const { Submission, SUBMISSION_JSON_SCHEMA } = require('./schemas');
const { validateDraft, buildToolResultRetry } = require('./validator');
const { buildContentBlocks } = require('./multimodal');
const { CACHEABLE_SYSTEM_BLOCKS, cacheableFewShot } = require('./prompts/system');

const TOOL_NAME = 'submit_extraction';
const TOOL_DESCRIPTION = 'Submit one farmOS draft with continuity decision and per-field confidence.';

function inlineTopLevelRef(schema) {
  // Plan 07 bug fix (Rule 1/3): zod-to-json-schema with a name arg emits
  // {$ref: '#/definitions/Foo', definitions: {Foo: {...}}}. Anthropic rejects
  // this with 400 "tools.0.custom.input_schema.type: Field required" because
  // input_schema must have `type` at the top level. Inline the named definition
  // while keeping `definitions` so nested $refs (e.g. for the discriminatedUnion
  // members) still resolve.
  if (!schema || typeof schema !== 'object' || !schema.$ref) return schema;
  const m = /^#\/definitions\/(.+)$/.exec(schema.$ref);
  if (!m || !schema.definitions || !schema.definitions[m[1]]) return schema;
  const inlined = { ...schema.definitions[m[1]], definitions: schema.definitions };
  delete inlined.$ref;
  return inlined;
}

function buildToolSpec() {
  return {
    name: TOOL_NAME,
    description: TOOL_DESCRIPTION,
    input_schema: inlineTopLevelRef(SUBMISSION_JSON_SCHEMA),
  };
}

function buildInitialUserContent({ captures, inFlightDraft, corpusContext, farmerCorrection }) {
  // Capture set -> a single user turn with: corpus context, in-flight summary, then per-capture text/transcript/images.
  // Plan 07 bug fix (Rule 1): close the last few-shot tool_use with a tool_result
  // block. Anthropic rejects 400 if any tool_use lacks an immediately-following
  // tool_result in the next user message.
  // Phase 53 BACK-03: last few-shot id is now tu_fewshot_6 (DT-tubs
  // physical_object_photo counter-example). tu_fewshot_3 and tu_fewshot_5 are
  // closed inline inside FEW_SHOT via leading tool_result blocks in the next
  // user turn (see prompts/system.js).
  const blocks = [];
  blocks.push({
    type: 'tool_result',
    tool_use_id: 'tu_fewshot_6',
    content: [{ type: 'text', text: 'accepted' }],
  });
  if (corpusContext && typeof corpusContext === 'object') {
    blocks.push({
      type: 'text',
      text: `corpus_context: ${JSON.stringify(corpusContext)}`,
    });
  }
  blocks.push({
    type: 'text',
    text: `In-flight draft: ${inFlightDraft ? JSON.stringify(inFlightDraft) : 'none'}`,
  });
  // Phase 39 D-03: append farmer-correction context only when present and non-empty.
  if (typeof farmerCorrection === 'string' && farmerCorrection.trim() !== '') {
    blocks.push({
      type: 'text',
      text: `Farmer correction: ${farmerCorrection.trim()}`,
    });
  }
  const captureList = Array.isArray(captures) ? captures : [];
  for (const cap of captureList) {
    const sub = buildContentBlocks({
      text: cap && cap.text,
      transcript: cap && cap.transcript,
      images: cap && cap.images,
    });
    for (const b of sub) blocks.push(b);
  }
  return blocks;
}

function findToolUseBlock(msg) {
  if (!msg || !Array.isArray(msg.content)) return null;
  return msg.content.find((b) => b && b.type === 'tool_use' && b.name === TOOL_NAME) || null;
}

function createExtractor({
  apiKey,
  logger = console,
  model = 'claude-sonnet-4-6',
  // Plan 09 (2026-05-12): default bumped 2048 -> 16384. Multi-event paper-log
  // pages produce 10-25 drafts; each ~150-250 output tokens; at 2048 the model
  // silently truncates the tool_use input to {} and the validator returns
  // schema_invalid (no drafts/continuity/continuity_reason). The original
  // 2048 default came from single-event Plan 03 sizing -- Plan 07 missed this
  // because the 2-fixture cap hit pages small enough to fit. Plan 08's
  // eval-two.js seed already used 16384 for the same reason.
  maxTokens = 16384,
  client: injectedClient = null,
  // Phase 54 Plan 03: optional observer fired once per Anthropic call. Live
  // capture sites omit this; only the backfill harness installs it to persist
  // every paid response to <runDir>/responses.jsonl. Errors in the observer
  // are caught + logged warn; they never propagate out of extract().
  onLlmCall = null,
} = {}) {
  const client = injectedClient || new Anthropic({ apiKey, maxRetries: 2 });
  const toolSpec = buildToolSpec();

  // Wrap client.messages.create so the observer fires for both the initial
  // call and the schema-retry call without duplicating the try/catch logic.
  async function callWithObserver(req, captureId) {
    const t0 = Date.now();
    let resp;
    let err;
    try {
      resp = await client.messages.create(req);
    } catch (e) {
      err = e;
      throw e;
    } finally {
      if (typeof onLlmCall === 'function') {
        try {
          const usage = (resp && resp.usage) || {};
          const requestHash = crypto.createHash('sha256')
            .update(JSON.stringify({
              model: req.model,
              messages: req.messages,
              tools: req.tools,
              system: req.system,
            }))
            .digest('hex')
            .slice(0, 16);
          await onLlmCall({
            ts: new Date().toISOString(),
            model: req.model,
            input_tokens: usage.input_tokens || 0,
            output_tokens: usage.output_tokens || 0,
            cache_creation_input_tokens: usage.cache_creation_input_tokens || 0,
            cache_read_input_tokens: usage.cache_read_input_tokens || 0,
            latency_ms: Date.now() - t0,
            raw_response: resp || null,
            request_hash: requestHash,
            captureId: captureId || null,
            error: err ? (err.message || String(err)) : null,
          });
        } catch (obsErr) {
          logger.warn && logger.warn(`[extractor] onLlmCall observer threw: ${obsErr.message}`);
        }
      }
    }
    return resp;
  }

  return {
    async extract({ captures, inFlightDraft, corpusContext, farmerCorrection } = {}) {
      const captureId = Array.isArray(captures) && captures[0] && captures[0].captureId
        ? captures[0].captureId
        : null;
      try {
        const systemBlocks = CACHEABLE_SYSTEM_BLOCKS;
        const fewShot = cacheableFewShot();
        const userContent = buildInitialUserContent({ captures, inFlightDraft, corpusContext, farmerCorrection });
        const messages = [...fewShot, { role: 'user', content: userContent }];

        const baseReq = {
          model,
          max_tokens: maxTokens,
          system: systemBlocks,
          tools: [toolSpec],
          tool_choice: { type: 'tool', name: TOOL_NAME },
          messages,
        };

        let resp;
        try {
          resp = await callWithObserver(baseReq, captureId);
        } catch (e) {
          logger.warn && logger.warn(`[extractor] degraded: ${e.message}`);
          return { ok: false, reason: e.message };
        }

        const usage1 = resp && resp.usage ? resp.usage : null;

        let toolUse = findToolUseBlock(resp);
        if (!toolUse) {
          return { ok: false, reason: 'no_tool_use_in_response', usage: usage1 };
        }

        let parsed = validateDraft(toolUse.input, Submission);
        if (parsed.ok) {
          return packResult(parsed.draft, sumUsage([usage1]));
        }

        // Retry once with tool_result is_error=true.
        const assistantTurn = { role: 'assistant', content: resp.content };
        const retryUserTurn = {
          role: 'user',
          content: [buildToolResultRetry(toolUse.id, parsed.errors)],
        };
        const retryReq = {
          ...baseReq,
          messages: [...messages, assistantTurn, retryUserTurn],
        };

        let resp2;
        try {
          resp2 = await callWithObserver(retryReq, captureId);
        } catch (e) {
          logger.warn && logger.warn(`[extractor] retry degraded: ${e.message}`);
          return { ok: false, reason: e.message, usage: usage1 };
        }

        const usage2 = resp2 && resp2.usage ? resp2.usage : null;
        const usageSum = sumUsage([usage1, usage2]);

        const toolUse2 = findToolUseBlock(resp2);
        if (!toolUse2) {
          return { ok: false, reason: 'no_tool_use_in_response', usage: usageSum };
        }
        const parsed2 = validateDraft(toolUse2.input, Submission);
        if (parsed2.ok) {
          return packResult(parsed2.draft, usageSum);
        }
        logger.warn && logger.warn(`[extractor] schema_invalid after retry`);
        return { ok: false, reason: 'schema_invalid', errors: parsed2.errors, raw_first: toolUse && toolUse.input, raw_retry: toolUse2 && toolUse2.input, usage: usageSum };
      } catch (e) {
        logger.warn && logger.warn(`[extractor] degraded: ${e.message}`);
        return { ok: false, reason: e.message };
      }
    },
  };
}

function packResult(submission, usage) {
  // Plan 08: multi-draft shape. Submission = {drafts: [{draft, per_field_confidence}], continuity, continuity_reason}.
  // Back-compat: also expose .draft/.per_field_confidence as the first element when drafts.length === 1
  // so call sites that still want the single-event view keep working until pipeline.js / state-machine.js are rewired.
  const drafts = Array.isArray(submission.drafts) ? submission.drafts : [];
  const first = drafts[0] || null;
  return {
    ok: true,
    drafts,
    continuity_decision: submission.continuity,
    continuity_reason: submission.continuity_reason,
    // Legacy fields (single-event view).
    draft: first ? first.draft : null,
    per_field_confidence: first ? first.per_field_confidence : null,
    // Phase 53 BACK-03: optional capture_kind passes through to callers
    // (analytics + future routing refinement). Omitted -> null. Routing
    // (BACK-02) does NOT consume this today -- back-compat lock.
    capture_kind: submission.capture_kind != null ? submission.capture_kind : null,
    usage: usage || null,
  };
}

function sumUsage(list) {
  const out = { input_tokens: 0, output_tokens: 0, cache_creation_input_tokens: 0, cache_read_input_tokens: 0 };
  let any = false;
  for (const u of list || []) {
    if (!u) continue;
    any = true;
    out.input_tokens += u.input_tokens || 0;
    out.output_tokens += u.output_tokens || 0;
    out.cache_creation_input_tokens += u.cache_creation_input_tokens || 0;
    out.cache_read_input_tokens += u.cache_read_input_tokens || 0;
  }
  return any ? out : null;
}

module.exports = {
  createExtractor,
  _internal: {
    buildInitialUserContent,
    findToolUseBlock,
    buildToolSpec,
    TOOL_NAME,
  },
};

File: src/farm-agent/farm_agent/extraction/extractor.py
"""
extraction/extractor.py -- Never-throws Sonnet extractor factory.

Port of src/agents/alerter/src/extraction/extractor.js (createExtractor).

Provides:
  create_extractor(client, model, max_tokens, on_llm_call, log) -> {"extract": extract}

Design decisions (from 60-CONTEXT.md):
  Area-1: Forced tool submit_extraction -- prevents free-form prose responses.
  Area-3: Fail-open on any error -- extractor never throws; every failure path returns
          {ok:False, reason} so the pipeline can produce a needs_review draft.
  Pitfall-9: with_options(timeout=_timeout_s) -- timeout is a transport option, NOT a
          body param (same live-fire bug Node hit; body kwarg -> 400).
  Pitfall-8: tu_fewshot_6 closer -- the first user-turn block MUST be a tool_result
          closing the last few-shot assistant tool_use. Anthropic 400s without it.
  Pitfall-1: block.id (NOT block.name) is used in tool_result retry turn's tool_use_id.
  Pitfall-5: second ValidationError must NOT propagate; catch and return {ok:False,...}.

Security:
  T-44-04-01: farmer text/transcript/image go into messages[] ONLY -- never concatenated
              into CACHEABLE_SYSTEM_BLOCKS (system prompt stays static).
  T-60-03-01: extra='forbid' on every nested model; invalid tool_use.input triggers retry
              then {ok:False}; never trusted raw.
  T-56-06-01: injected client owns the api_key; extractor never references or logs it.
  T-59-02-01: WARNING logs contain only exception/reason strings -- never farmer content.
  T-60-03-02: per-request timeout via client.with_options(timeout=...).
"""

from __future__ import annotations

import asyncio
import datetime as _dt
import decimal
import inspect
import json
import logging
import uuid

import anthropic
from pydantic import ValidationError

from farm_agent.extraction.multimodal import build_content_blocks
from farm_agent.extraction.prompts import CACHEABLE_SYSTEM_BLOCKS, cacheable_few_shot
from farm_agent.extraction.schemas.submission import SUBMISSION_JSON_SCHEMA, Submission

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Module-level constants
# ---------------------------------------------------------------------------

TOOL_NAME = "submit_extraction"
TOOL_DESCRIPTION = (
    "Submit the structured extraction result for this farm operation capture. "
    "Call this tool exactly once with the complete structured draft."
)

# Timeout mirrors the Node extractor (60 000ms) -- longer than the gate (2 000ms)
# because multi-event seeding sessions can produce large responses.
_DEFAULT_TIMEOUT_MS = 60_000


# ---------------------------------------------------------------------------
# JSON serialization of raw DB rows (MUSHY-76)
# ---------------------------------------------------------------------------


def json_default(o):
    """`default=` hook for json.dumps over raw psycopg rows.

    The in-flight draft handed to build_initial_user_content() is a RAW
    signal_draft row (extraction_db.get_in_flight_for_sender zips
    cursor.description over the tuple), so created_at / updated_at /
    confirmed_at / expired_at / nudge_sent_at arrive as python datetimes and
    plain json.dumps raises TypeError before the model is ever called.

    Node is the shape source of truth: `pg` hands JSON.stringify JS Dates
    (-> ISO-8601 strings), numerics and uuids as strings. This mirrors that:

      datetime / date / time -> .isoformat()   (matches JS Date -> ISO-8601)
      Decimal                -> str            (matches pg numeric -> JS string)
      UUID                   -> str            (matches pg uuid  -> JS string)

    DELIBERATELY NOT HANDLED -- these raise TypeError rather than being coerced:
      bytes / memoryview (bytea), timedelta (interval), sets, model objects.
    A `default=str` catch-all would silently turn structured values into
    garbage prose that the model then reasons over as if it were data. There is
    no bytea or interval column on signal_draft, so a raise here means a real
    schema change nobody accounted for -- fail loudly (the extractor's outer
    fail-open still turns it into {ok:False}) rather than quietly mislead.
    """
    if isinstance(o, (_dt.datetime, _dt.date, _dt.time)):
        return o.isoformat()
    if isinstance(o, decimal.Decimal):
        return str(o)
    if isinstance(o, uuid.UUID):
        return str(o)
    raise TypeError(f"Object of type {type(o).__name__} is not JSON serializable")


def _dumps(obj) -> str:
    """json.dumps with the row-aware default hook. Use at every dumps site here."""
    return json.dumps(obj, default=json_default)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def build_tool_spec() -> dict:
    """Build the tool spec dict for the Anthropic messages API.

    Passes SUBMISSION_JSON_SCHEMA DIRECTLY as input_schema.  Pydantic v2 already
    emits type:object at the root -- no inlineTopLevelRef wrapper needed (unlike
    the Node zod-to-json-schema approach).
    """
    return {
        "name": TOOL_NAME,
        "description": TOOL_DESCRIPTION,
        "input_schema": SUBMISSION_JSON_SCHEMA,
    }


def find_tool_use_block(response) -> object | None:
    """Return the first submit_extraction tool_use block, or None.

    Pitfall 7: block.type / block.name / block.id / block.input are ATTRIBUTE
    access on pydantic response objects, NOT dict keys.
    """
    if not response or not hasattr(response, "content"):
        return None
    for block in response.content:
        if block.type == "tool_use" and block.name == TOOL_NAME:
            return block
    return None


def build_initial_user_content(
    captures: list[dict],
    in_flight_draft=None,
    corpus_context=None,
    farmer_correction: str | None = None,
) -> list[dict]:
    """Build the user message content list for the extraction call.

    Pitfall 8: The FIRST block MUST be a tool_result closing tu_fewshot_6.
    Anthropic 400s when any tool_use in an assistant turn lacks a following
    tool_result in the next user turn.

    Security T-44-04-01: farmer text/transcript/image enter via messages[] only.
    Never concatenated into CACHEABLE_SYSTEM_BLOCKS.
    """
    blocks: list[dict] = []

    # Mandatory: close the last few-shot assistant tool_use (tu_fewshot_6)
    blocks.append({
        "type": "tool_result",
        "tool_use_id": "tu_fewshot_6",
        "content": [{"type": "text", "text": "accepted"}],
    })

    # Optional corpus context block (isinstance guard mirrors Node typeof === 'object')
    if corpus_context is not None and isinstance(corpus_context, dict):
        blocks.append({
            "type": "text",
            "text": f"corpus_context: {_dumps(corpus_context)}",
        })

    # In-flight draft block (always present; "none" when absent)
    blocks.append({
        "type": "text",
        "text": f"In-flight draft: {_dumps(in_flight_draft) if in_flight_draft is not None else 'none'}",
    })

    # Optional farmer correction block
    if farmer_correction and farmer_correction.strip():
        blocks.append({
            "type": "text",
            "text": f"Farmer correction: {farmer_correction.strip()}",
        })

    # Per-capture content blocks (text, transcript, image blocks)
    for cap in captures:
        for block in build_content_blocks(
            text=cap.get("text"),
            transcript=cap.get("transcript"),
            images=cap.get("images") or [],
        ):
            blocks.append(block)

    return blocks


def sum_usage(usages: list) -> dict | None:
    """Sum token counts across multiple usage objects (initial + retry calls).

    Mirrors Node sumUsage: returns None when ALL usages in the list are null
    (tracks an `any_data` flag). Returns a zeroed dict only when at least one
    non-null usage was present -- so downstream `result["usage"] is None` checks
    correctly detect "no usage data recorded" on failure/degraded paths.

    Uses attribute access with defaults of 0 for missing fields.
    """
    total: dict[str, int] = {
        "input_tokens": 0,
        "output_tokens": 0,
        "cache_creation_input_tokens": 0,
        "cache_read_input_tokens": 0,
    }
    any_data = False
    for u in usages:
        if u is None:
            continue
        any_data = True
        for key in total:
            total[key] += getattr(u, key, 0)
    return total if any_data else None


def pack_result(submission: Submission, usage: dict) -> dict:
    """Expand a validated Submission into the canonical extractor result dict.

    Mirrors Node's packResult (extractor.js lines 250-270).

    Node's packResult returns the zod-validated tool input as PLAIN JS objects, and
    every consumer in this port (pipeline, batch_mode, starting_seq, preview_builder,
    state_machine, seq_helper) was written against that shape -- they do dict access
    (draft.get("type"), draft["groups"]).  So the pydantic models MUST be dumped to
    plain python data here, at the boundary, not handled downstream.

    mode="json" so nested models, enums and any date-like fields come out as
    JSON-safe primitives: draft_json is a jsonb column written via
    psycopg.types.json.Jsonb(...), which cannot adapt a BaseModel either.
    """
    dumped = submission.model_dump(mode="json")
    drafts: list[dict] = list(dumped.get("drafts") or [])
    first = drafts[0] if drafts else None
    return {
        "ok": True,
        "drafts": drafts,
        "continuity_decision": dumped.get("continuity"),
        "continuity_reason": dumped.get("continuity_reason"),
        "draft": first.get("draft") if first is not None else None,
        "per_field_confidence": first.get("per_field_confidence") if first is not None else None,
        "capture_kind": dumped.get("capture_kind"),
        "usage": usage,
    }


# ---------------------------------------------------------------------------
# Factory
# ---------------------------------------------------------------------------


def create_extractor(
    client: anthropic.AsyncAnthropic,
    model: str = "claude-sonnet-4-6",
    max_tokens: int = 16384,
    on_llm_call=None,
    log: logging.Logger | None = None,
) -> dict:
    """Factory returning {"extract": extract}. Port of createExtractor().

    Holds the injected AsyncAnthropic in the closure. Do NOT create a second
    client per call; share the single daemon-lifetime singleton.

    Args:
        client:      Injected AsyncAnthropic (api_key managed externally, never logged).
        model:       Sonnet model ID (verbatim from Node extraction extractor).
        max_tokens:  Maximum tokens in response (16384 for multi-event pages).
        on_llm_call: Optional observer called after each LLM call with (req, resp).
                     Accepts both sync and async callables. Observer errors are
                     caught + warn-logged, never propagated.
        log:         Optional logger; defaults to module logger.

    Returns:
        {"extract": async_fn} where async_fn never raises.
    """
    _log = log or logger
    _timeout_s = _DEFAULT_TIMEOUT_MS / 1000

    async def _call_with_observer(req: dict, _capture_id: str | None = None):
        """Wrap client.with_options(timeout=...).messages.create(**req).

        Fires on_llm_call observer in a finally block (both initial and retry calls),
        mirroring Node's callWithObserver which fires onLlmCall in finally so timed-out
        or errored calls that consumed tokens are still recorded. Observer receives an
        `error` field (str|None) so the backfill harness can log partial-response costs.

        Observer errors are caught + warn-logged, never propagated (mirror Node).
        Uses inspect.iscoroutinefunction to decide await, not iscoroutine(result).
        """
        resp = None
        exc = None
        try:
            # Pitfall 9: timeout via with_options, NEVER as a body kwarg
            resp = await client.with_options(timeout=_timeout_s).messages.create(**req)
            return resp
        except Exception as e:  # noqa: BLE001
            exc = e
            raise
        finally:
            if on_llm_call is not None:
                try:
                    if inspect.iscoroutinefunction(on_llm_call):
                        await on_llm_call(req, resp, exc)
                    else:
                        on_llm_call(req, resp, exc)
                except Exception as obs_err:  # noqa: BLE001
                    _log.warning("[extractor] observer error: %s", obs_err)

    async def extract(
        captures: list[dict],
        in_flight_draft=None,
        corpus_context=None,
        farmer_correction: str | None = None,
    ) -> dict:
        """Extract a structured draft from captures via Sonnet forced tool-use.

        Never raises (fail-open). Every failure path returns {ok:False, reason:...}.

        Returns (NEVER raises):
          {ok:True, drafts, continuity_decision, continuity_reason, draft,
           per_field_confidence, capture_kind, usage}  on success
          {ok:False, reason:str | "no_tool_use_in_response" | "schema_invalid",
           ...}  on any failure

        Security T-59-02-01: farmer content MUST NOT appear in log output.
        """
        try:
            user_content = build_initial_user_content(
                captures=captures,
                in_flight_draft=in_flight_draft,
                corpus_context=corpus_context,
                farmer_correction=farmer_correction,
            )
            messages = [
                *cacheable_few_shot(),
                {"role": "user", "content": user_content},
            ]
            base_req: dict = {
                "model": model,
                "max_tokens": max_tokens,
                "system": CACHEABLE_SYSTEM_BLOCKS,
                "tools": [build_tool_spec()],
                "tool_choice": {"type": "tool", "name": TOOL_NAME},
                "messages": messages,
            }

            # ---------------------------------------------------------------
            # First LLM call
            # ---------------------------------------------------------------
            try:
                resp = await _call_with_observer(base_req)
            except Exception as e:  # noqa: BLE001
                _log.warning("[extractor] first call degraded: %s", e)
                return {"ok": False, "reason": str(e)}

            block = find_tool_use_block(resp)
            if block is None:
                return {"ok": False, "reason": "no_tool_use_in_response"}

            # ---------------------------------------------------------------
            # Validate first response
            # ---------------------------------------------------------------
            first_validation_error: str | None = None
            try:
                submission = Submission.model_validate(block.input)
                return pack_result(submission, sum_usage([getattr(resp, "usage", None)]))
            except ValidationError as ve:
                first_validation_error = str(ve)
                _log.warning("[extractor] first validation failed, retrying: %s", ve)

            # ---------------------------------------------------------------
            # Retry: one tool_result is_error=True turn (Pitfall 1: block.id not block.name)
            # ---------------------------------------------------------------
            assistant_turn = {"role": "assistant", "content": resp.content}
            retry_user_turn = {
                "role": "user",
                "content": [
                    {
                        "type": "tool_result",
                        "tool_use_id": block.id,  # block.id, NOT block.name (Pitfall 1)
                        "is_error": True,
                        "content": first_validation_error,
                    }
                ],
            }
            retry_req = {
                **base_req,
                "messages": [*messages, assistant_turn, retry_user_turn],
            }

            try:
                resp2 = await _call_with_observer(retry_req)
            except Exception as e2:  # noqa: BLE001
                _log.warning("[extractor] retry call degraded: %s", e2)
                return {"ok": False, "reason": str(e2)}

            block2 = find_tool_use_block(resp2)
            if block2 is None:
                return {"ok": False, "reason": "no_tool_use_in_response"}

            # ---------------------------------------------------------------
            # Validate retry response (Pitfall 5: MUST catch, never propagate)
            # ---------------------------------------------------------------
            try:
                submission2 = Submission.model_validate(block2.input)
                combined_usage = sum_usage([
                    getattr(resp, "usage", None),
                    getattr(resp2, "usage", None),
                ])
                return pack_result(submission2, combined_usage)
            except ValidationError as e2:  # noqa: BLE001 -- Pitfall 5: must not propagate
                return {
                    "ok": False,
                    "reason": "schema_invalid",
                    "errors": e2.errors(),
                    "raw_first": block.input,
                    "raw_retry": block2.input,
                }

        except Exception as e:  # noqa: BLE001 -- outer fail-open guard
            _log.warning("[extractor] degraded: %s", e)
            return {"ok": False, "reason": str(e)}

    return {"extract": extract}
